### PR TITLE
Recreate VACOLS full grant and remand reports for EP

### LINKED ISSUE #352: Recreate AMC full grant report from VACOLS
Obtain details from Jed, and recreate the report in Ruby/SQL.

### PR DESCRIPTION
**Pull request by cmgiven**
fixes #352.

@shanear @joofsh: Mind taking a look at this? Will this dovetail with what you're planning?

One interesting discovery: not everything in remand status (`BFMPRO = 'REM'`) is marked as a remand by disposition (`BFDC`). Unfortunately, that means that the approach here, discussed with Shane, of `AppealRepository.ready_for_end_product` returning both remands and full grants, means that we don't currently have a way to distinguish the two. We'll need to add more information to the `appeal` model to do so. Hanging onto `BFMPRO` would do the trick. Or we could abandon concatenating remands and full grants together into a single method.

### CODE REVIEW

**Comment by joofsh** [comment]
I wasn't part of the original conversation, but is there a particular reason we want to combine these into a single method?

It seems awkward that we merge them into one method, especially since one query is date scoped and the other not. If we want to combine these into a unified work queue, we can still execute both methods in the same background job and merge the task arrays in the UI.

File: app/models/vacols/case.rb [hunk 1]
```diff
@@ -18,6 +18,30 @@ class VACOLS::Case < VACOLS::Record
...
+    "F" => "Advance Withdrawn by Appellant/Rep",
+    "G" => "Advance Failure to Respond",
+    "L" => "Manlincon Remand",
+    "M" => "Merged Appeal",
+    "Q" => "Recon Motion Withdrawn",
+    "R" => "Reconsideration by Letter",
+    "V" => "Motion to Vacate Withdrawn",
+    "W" => "Withdrawn from Remand",
+    "X" => "Remand Failure to Respond",
+    nil => nil
```
Comment (joofsh) [inline app/models/vacols/case.rb]: Minor nit: you don't need to add the `nil => nil`

File: app/models/vacols/case.rb [hunk 2]
```diff
@@ -60,4 +84,37 @@ class VACOLS::Case < VACOLS::Record
...
+      BFMPRO = 'REM'
+      -- Remand status.
+
+      and BFCURLOC = '97'
+      -- Currently sitting in loc 97.
+
+    ")
+  end
+
+  def self.fullgrants_for_ep(starttime)
```
Comment (joofsh) [inline app/models/vacols/case.rb]: Very subjective (and I don't have a strong opinion), but maybe we could use a more descriptive param here? something like `granted_after`? I wouldn't know what `starttime` means without more comments

**Comment by joofsh** [comment]
Added some minor feedback, but to answer your original question, yes! This looks great and will definitely be useful for the dispatch Work Queue 👍  🙌

File: app/models/vacols/case.rb (hunk 1 above)
Comment (cmgiven) [inline app/models/vacols/case.rb]: Yup. Was originally mapping to a hash with multiple values. Opted to cut until actually needed.

File: app/models/vacols/case.rb (hunk 2 above)
Comment (cmgiven) [inline app/models/vacols/case.rb]: Makes sense.

File: app/models/vacols/case.rb [hunk 3]
```diff
@@ -60,4 +83,37 @@ class VACOLS::Case < VACOLS::Record
...
+      BFMPRO = 'REM'
+      -- Remand status.
+
+      and BFCURLOC = '97'
+      -- Currently sitting in loc 97.
+
+    ")
+  end
+
+  def self.fullgrants_for_ep(decided_after)
```
Comment (shanear) [inline app/models/vacols/case.rb]: Why is `fullgrant` one word?

File: app/services/fullgrants_report.rb [hunk 4]
```diff
@@ -0,0 +1,42 @@
+class FullgrantsReport < Report
```
Comment (shanear) [inline app/services/fullgrants_report.rb]: will we end up needing these reports after the work queue is in place?

File: app/services/appeal_repository.rb [hunk 5]
```diff
@@ -25,6 +25,22 @@ def self.find(vacols_id, _args = {})
...
 
+  def self.remands_for_end_product
+    remands = MetricsService.timer "loaded remands in loc 97 from VACOLS" do
+      VACOLS::CASE.remands_for_ep
+    end
+
+    remands.map { |case_record| create_appeal(case_record) }
+  end
+
+  def self.fullgrants_for_end_product(decided_after)
```
Comment (shanear) [inline app/services/appeal_repository.rb]: why do we spell out `end_product` on this method and say `ep` on `VACOLS::Case`?

Comment (shanear) [inline app/services/appeal_repository.rb]: also, `fullgrants_for_end_product` feels a little unclear to me. maybe `fullgrants_requiring_end_product` or `fullgrants_needing_end_product` or `fullgrants_waiting_for_end_product`?

File: app/services/fullgrants_report.rb (hunk 4 above)
Comment (cmgiven) [inline app/services/fullgrants_report.rb]: No. Just wanted to make it easy to run and check what's appearing in each.

File: app/services/appeal_repository.rb (hunk 5 above)
Comment (cmgiven) [inline app/services/appeal_repository.rb]: Per @abbyraskin, we may want to move off of using "end product" as the key term here. It's an artifact of the current process, and not used universally (some ROs do not create EPs, but track their work separately). The term used currently is "claims establishment" (CEST). Which is a mouthful, but probably good for consistency.

File: app/models/vacols/case.rb [hunk 6]
```diff
@@ -60,4 +83,37 @@ class VACOLS::Case < VACOLS::Record
...
+      BFMPRO = 'REM'
+      -- Remand status.
+
+      and BFCURLOC = '97'
+      -- Currently sitting in loc 97.
+
+    ")
+  end
+
+  def self.full_grants_for_claims_establishment(decided_after)
```
Comment (shanear) [inline app/models/vacols/case.rb]: per our discussion, change to `amc_full_grants(decided_after:)`

File: app/services/appeal_repository.rb [hunk 7]
```diff
@@ -25,6 +25,22 @@ def self.find(vacols_id, _args = {})
...
 
+  def self.remands_for_claims_establishment
+    remands = MetricsService.timer "loaded remands in loc 97 from VACOLS" do
+      VACOLS::CASE.remands_for_claims_establishment
+    end
+
+    remands.map { |case_record| create_appeal(case_record) }
+  end
+
+  def self.full_grants_for_claims_establishment(decided_after)
```
Comment (shanear) [inline app/services/appeal_repository.rb]: `amc_full_grants(decided_after:)`

**Pull request merged**